rmf_websocket/src/rmf_websocket/client/ClientWebSocketEndpoint.cpp

```diff
@@ -0,0 +1,235 @@
+  _endpoint.init_asio(io_service);
+  _endpoint.start_perpetual();
+}
+
+//=============================================================================
+websocketpp::lib::error_code ClientWebSocketEndpoint::connect()
+{
+  websocketpp::lib::error_code ec;
+
+  _init = true;
+  _con = _endpoint.get_connection(_uri, ec);
+
+  if (ec)
+  {
+    std::stringstream err;
+    err << "> Connect initialization error: " << ec.message();
+    _logger(err.str());
+    return ec;
+  }
+
+  auto reconnect_socket = [this]()
+    {
+      // TODO(arjo) Parametrize the timeout.
+      using namespace std::chrono_literals;
+      _logger("Attempting reconnection in 1s.");
+      std::this_thread::sleep_for(1s);
+      connect();
+    };
+
+  // Not sure why but seems like I have to re-initallize this everytime in order
+  // to actually make a clean connection. My guess is the shared pointer is being
+  // leaked somewhere.
+  _current_connection = std::make_shared<ConnectionMetadata>(
+    _con->get_handle(),
+    _uri, _connection_cb, reconnect_socket);
+
+  _con->set_open_handler(websocketpp::lib::bind(
+      &ConnectionMetadata::on_open,
+      _current_connection,
+      &_endpoint,
+      websocketpp::lib::placeholders::_1
+  ));
+  _con->set_fail_handler(websocketpp::lib::bind(
+      &ConnectionMetadata::on_fail,
+      _current_connection,
+      &_endpoint,
+      websocketpp::lib::placeholders::_1
+  ));
+
+  _con->set_close_handler(websocketpp::lib::bind(
+      &ConnectionMetadata::on_close,
+      _current_connection,
+      &_endpoint,
+      websocketpp::lib::placeholders::_1
+  ));
+
+
+  _endpoint.connect(_con);
+
+  return ec;
+}
+
+//=============================================================================
+std::optional<ConnectionMetadata::ConnectionStatus> ClientWebSocketEndpoint::
+get_status() const
+{
+  if (!_init)
+  {
+    return std::nullopt;
+  }
+  return _current_connection->get_status();
+}
+
+//=============================================================================
+websocketpp::lib::error_code ClientWebSocketEndpoint::send(
+  const std::string& message)
+{
+  websocketpp::lib::error_code ec;
+
+  _endpoint.send(
+    _current_connection->get_hdl(), message, websocketpp::frame::opcode::text,
+    ec);
+  if (ec)
+  {
+    return ec;
+  }
+  return ec;
+}
+
+//=============================================================================
+ClientWebSocketEndpoint::~ClientWebSocketEndpoint()
+{
+  _endpoint.stop_perpetual();
+
+  if (!_current_connection)
+  {
+    return;
+  }
+
+  if (_current_connection->get_status() !=
+    ConnectionMetadata::ConnectionStatus::OPEN)
+  {
+    // Only close open connections
+    return;
+  }
+
+
+  websocketpp::lib::error_code ec;
+  _endpoint.close(
+    _current_connection->get_hdl(), websocketpp::close::status::going_away, "",
+    ec);
+  if (ec)
+  {
+    std::stringstream err;
+    err << "> Error closing connection : " << ec.message();
+    _logger(err.str());
```

Comment:
Can we also log the URI that we are closing?

Reply:
29ddac49f6a2813fde9992bc60a19057171c39c0